File: app/src/main/java/com/aykuttasil/sweetloc/activity/MainActivity.java
package com.aykuttasil.sweetloc.activity;

import android.content.Intent;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.FrameLayout;

import com.aykuttasil.sweetloc.R;
import com.aykuttasil.sweetloc.db.DbManager;
import com.aykuttasil.sweetloc.fragment.UserTrackerListFragment_;
import com.aykuttasil.sweetloc.helper.SuperHelper;
import com.aykuttasil.sweetloc.model.ModelUser;
import com.onesignal.OneSignal;
import com.orhanobut.logger.Logger;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.EActivity;
import org.androidannotations.annotations.OnActivityResult;
import org.androidannotations.annotations.ViewById;

import hugo.weaving.DebugLog;
import rx.Observable;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

/**
 * Created by aykutasil on 23.06.2016.
 */
@EActivity(R.layout.activity_main)
public class MainActivity extends BaseActivity {

    @ViewById(R.id.Toolbar)
    Toolbar mToolbar;

    @ViewById(R.id.Container)
    FrameLayout mContainer;

    //

    @DebugLog
    @AfterViews
    public void initializeAfterViews() {
        initToolbar();
        updateUi();
    }

    @DebugLog
    @Override
    void initToolbar() {
        setSupportActionBar(mToolbar);
        getSupportActionBar().setTitle("SweetLoc");
    }

    @DebugLog
    @Override
    void updateUi() {

        if (!SuperHelper.checkUser()) {
            SuperHelper.logoutUser();
            goLoginFacebookActivity(this);
        } else {
            initMain();
        }
    }

    @DebugLog
    private void initMain() {

        SuperHelper.startPeriodicTask(this);

        if (DbManager.getOneSignalUserId() == null) {

            Observable.create(
                    new Observable.OnSubscribe<String>() {
                        @Override
                        public void call(Subscriber<? super String> subscriber) {

                            OneSignal.idsAvailable((userId, registrationId) -> {

                                Logger.i("OneSignal userId: " + userId);
                                Logger.i("OneSignal regId: " + registrationId);

                                subscriber.onNext(userId);

                            });
                        }
                    })
                    .flatMap(new Func1<String, Observable<String>>() {
                        @Override
                        public Observable<String> call(String userId) {

                            ModelUser modelUser = DbManager.getModelUser();
                            modelUser.setOneSignalUserId(userId);
                            modelUser.save();

                            SuperHelper.updateUser(modelUser);

                            return Observable.just(userId);
                        }
                    })
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(result -> {
                        goFragment();
                    }, SuperHelper::CrashlyticsError);
        } else {
            goFragment();
        }

    }

    @DebugLog
    private void goFragment() {

        SuperHelper.ReplaceFragmentBeginTransaction(
                MainActivity.this,
                UserTrackerListFragment_.builder().build(),
                R.id.Container,
                false);
    }

    /**
     * Activity singleTop modunda çalıştırılır ise ve
     * bu Activity zaten görünür durumda ise
     * yani stack in en üstünde ise startActivity() ile bu activity ye gönderilen intent buraya ulaşır.
     * Activity nin yeni bir instance ı oluşturulmaz.
     *
     * @param intent
     */
    @DebugLog
    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

        updateUi();
    }

    @DebugLog
    @OnActivityResult(LOGIN_REQUEST_CODE)
    public void ActivityResultLogin(int resultCode) {

        switch (resultCode) {
            case RESULT_OK: {
                updateUi();
                break;
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        getMenuInflater().inflate(R.menu.menu_mainactivity, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @DebugLog
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {
            case R.id.menuProfil: {
                ProfileActivity_.intent(this).flags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK).start();
                break;
            }
            case R.id.menuMap: {
                MapsActivity_.intent(this).flags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK).start();
                break;
            }
        }
        return true;
    }


    @DebugLog
    @Override
    protected void onStop() {
        super.onStop();
    }

    @DebugLog
    @Override
    protected void onDestroy() {
        super.onDestroy();
    }
}

File: app/src/main/java/com/aykuttasil/sweetloc/activity/main/MainActivity.kt
package com.aykuttasil.sweetloc.activity.main

import android.annotation.TargetApi
import android.app.Activity
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.os.Build
import android.view.Menu
import android.view.MenuItem
import com.aykuttasil.sweetloc.R
import com.aykuttasil.sweetloc.activity.base.BaseActivity
import com.aykuttasil.sweetloc.activity.login.LoginActivity_
import com.aykuttasil.sweetloc.activity.map.MapsActivity_
import com.aykuttasil.sweetloc.activity.profile.ProfileActivity_
import com.aykuttasil.sweetloc.db.DbManager
import com.aykuttasil.sweetloc.fragment.UserTrackerListFragment_
import com.aykuttasil.sweetloc.helper.SuperHelper
import com.onesignal.OneSignal
import com.orhanobut.logger.Logger
import hugo.weaving.DebugLog
import kotlinx.android.synthetic.main.activity_main.*
import org.androidannotations.annotations.AfterViews
import org.androidannotations.annotations.EActivity
import org.androidannotations.annotations.OnActivityResult

/**
 * Created by aykutasil on 23.06.2016.
 */
@EActivity(R.layout.activity_main)
open class MainActivity : BaseActivity() {

    @DebugLog
    @AfterViews
    override fun initializeAfterViews() {
        initToolbar()
        updateUi()
    }

    @DebugLog
    override fun initToolbar() {
        setSupportActionBar(Toolbar)
        supportActionBar?.title = "SweetLoc"
    }

    @DebugLog
    override fun updateUi() {
        if (!SuperHelper.checkUser()) {
            SuperHelper.logoutUser()
            LoginActivity_.intent(this).startForResult(BaseActivity.LOGIN_REQUEST_CODE)
            // goLoginFacebookActivity(this);
        } else {
            initMain()
        }
    }

    @DebugLog
    private fun initMain() {
        SuperHelper.startPeriodicTask(this)
        checkAndOpenAudioSettings()

        if (DbManager.getOneSignalUserId() == null) {
            OneSignal.idsAvailable { userId, registrationId ->
                Logger.i("OneSignal userId: " + userId)
                Logger.i("OneSignal regId: " + registrationId)

                val modelUser = DbManager.getModelUser()
                modelUser.oneSignalUserId = userId
                modelUser.save()
                SuperHelper.updateUser(modelUser)

                goFragment()
            }
        } else {
            goFragment()
        }
    }

    @DebugLog
    private fun goFragment() {
        SuperHelper.ReplaceFragmentBeginTransaction(
                this@MainActivity,
                UserTrackerListFragment_.builder().build(),
                R.id.Container,
                false)
    }

    @DebugLog
    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        updateUi()
    }

    @DebugLog
    @OnActivityResult(BaseActivity.LOGIN_REQUEST_CODE)
    fun activityResultLogin(resultCode: Int) {
        when (resultCode) {
            Activity.RESULT_OK -> {
                updateUi()
            }
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    private fun checkAndOpenAudioSettings() {
        val notificationManager = this.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && !notificationManager.isNotificationPolicyAccessGranted) {
            val intent = Intent(android.provider.Settings.ACTION_NOTIFICATION_POLICY_ACCESS_SETTINGS)
            startActivity(intent)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_mainactivity, menu)
        return super.onCreateOptionsMenu(menu)
    }

    @DebugLog
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menuProfil -> {
                ProfileActivity_.intent(this).start()
            }
            R.id.menuMap -> {
                MapsActivity_.intent(this).start()
            }
        }
        return true
    }
}
